File: dotNetTips.Utility/Services.vb
' ***********************************************************************
' Assembly         : dotNetTips.Utility
' Author           : david
' Created          : 05-05-2017
'
' Last Modified By : david
' Last Modified On : 07-18-2017
' ***********************************************************************
' <copyright file="Services.vb" company="McCarter Consulting - David McCarter">
'     David McCarter - dotNetTips.com © 2017
' </copyright>
' <summary></summary>
' *************************************************************************
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.ServiceProcess
Imports dotNetTips.Utility.Portable.OOP

''' <summary>
''' Class that deals with Windows services.
''' </summary>
Public Module Services

    ''' <summary>
    ''' Loads the service.
    ''' </summary>
    ''' <param name="serviceName">Name of the service.</param>
    ''' <returns>ServiceController.</returns>
    Private Function LoadService(serviceName As String) As ServiceController
        Return ServiceController.GetServices().FirstOrDefault(Function(p) p.ServiceName = serviceName)
    End Function

    ''' <summary>
    ''' Alls the services.
    ''' </summary>
    ''' <returns>IEnumerable(Of System.String).</returns>
    Public Function AllServices() As IEnumerable(Of String)
        Return ServiceController.GetServices().[Select](Function(p) p.ServiceName).AsEnumerable()
    End Function

    ''' <summary>
    ''' Services the exists.
    ''' </summary>
    ''' <param name="serviceName">Name of the service.</param>
    ''' <returns><c>true</c> if XXXX, <c>false</c> otherwise.</returns>
    Public Function ServiceExists(serviceName As String) As Boolean
        Encapsulation.TryValidateParam(serviceName, NameOf(serviceName))

        Dim service = LoadService(serviceName)

        Return If(service IsNot Nothing, True, False)
    End Function

    ''' <summary>
    ''' Gets the status of a Windows service.
    ''' </summary>
    ''' <param name="serviceName">Name of the service.</param>
    ''' <returns>ServiceControllerStatus.</returns>
    ''' <exception cref="InvalidOperationException">Service not found.</exception>
    ''' <exception cref="System.InvalidOperationException">Service not found.</exception>
    Public Function ServiceStatus(serviceName As String) As ServiceControllerStatus
        Encapsulation.TryValidateParam(serviceName, NameOf(serviceName))

        If ServiceExists(serviceName) Then
            Dim service = LoadService(serviceName)

            If service IsNot Nothing Then
                Return service.Status
            End If
        End If
    End Function

    ''' <summary>
    ''' Starts the service.
    ''' </summary>
    ''' <param name="serviceName">Name of the service.</param>
    ''' <returns>ServiceActionResult.</returns>
    Public Function StartService(serviceName As String) As ServiceActionResult
        Encapsulation.TryValidateParam(serviceName, NameOf(serviceName))

        Dim statusResult = ServiceActionResult.NotFound

        If ServiceExists(serviceName) = False Then
            Return statusResult
        End If

        Try
            Dim service = LoadService(serviceName)

            If (service IsNot Nothing AndAlso service.Status = ServiceControllerStatus.Stopped) Then
                service.Start()
                statusResult = ServiceActionResult.Running
            End If

        Catch win32Ex As Win32Exception
            statusResult = ServiceActionResult.Error
        Catch invalidEx As InvalidOperationException
            statusResult = ServiceActionResult.Error
        End Try

        Return statusResult
    End Function

    ''' <summary>
    ''' Starts the stop services.
    ''' </summary>
    ''' <param name="services">The requests.</param>
    Public Sub StopServices(services As IEnumerable(Of ServiceAction))
        Encapsulation.TryValidateParam(Of ArgumentNullException)(services IsNot Nothing, NameOf(services))

        For Each service In services.AsParallel
            service.ServiceActionResult = StopService(service.ServiceName)
        Next

    End Sub


    ''' <summary>
    ''' Starts the services.	
    ''' </summary>
    ''' <param name="services">The services.</param>
    ''' <remarks></remarks>
    Public Sub StartServices(services As IEnumerable(Of ServiceAction))
        Encapsulation.TryValidateParam(Of ArgumentNullException)(services IsNot Nothing, NameOf(services))

        For Each service In services.AsParallel
            service.ServiceActionResult = StartService(service.ServiceName)
        Next

    End Sub

    ''' <summary>
    ''' Get the status of a Windows Service.	
    ''' </summary>
    ''' <param name="services">The services.</param>
    ''' <remarks></remarks>
    Public Sub ServicesStatus(services As IEnumerable(Of ServiceAction))
        Encapsulation.TryValidateParam(Of ArgumentNullException)(services IsNot Nothing, NameOf(services))

        ''Find status on services
        For Each service In services.AsParallel
            service.ServiceActionResult = ServiceStatus(service.ServiceName)
        Next
    End Sub

    ''' <summary>
    ''' Stops the service.
    ''' </summary>
    ''' <param name="serviceName">Name of the service.</param>
    ''' <returns>ServiceActionResult.</returns>
    Public Function StopService(serviceName As String) As ServiceActionResult
        Encapsulation.TryValidateParam(serviceName, NameOf(serviceName))

        Dim statusResult As Object = ServiceActionResult.NotFound

        If ServiceExists(serviceName) = False Then
            Return statusResult
        End If

        Try
            Dim service = LoadService(serviceName)

            If (service IsNot Nothing AndAlso service.Status = ServiceControllerStatus.Running) Then
                service.Stop()
                statusResult = ServiceActionResult.Stopped
            End If
        Catch win32Ex As Win32Exception
            statusResult = ServiceActionResult.Error
        Catch invalidEx As InvalidOperationException
            statusResult = ServiceActionResult.Error
        End Try

        Return statusResult
    End Function
End Module

File: dotNetTips.Utility.Standard/Services.cs
using dotNetTips.Utility.Standard.OOP;
using System;
using System.Collections.Generic;
using System.Linq;
using System.ServiceProcess;
using System.Text;
using System.Collections;

namespace dotNetTips.Utility.Standard
{
    public static class Services
    {
        //TODO: BLOG POST
        public static ServiceControllerStatus ServiceStatus(string serviceName)
        {
            var service = LoadService(serviceName);

            if (service != null)
            {
                return service.Status;
            }
            else
            {
                throw new InvalidOperationException("Service not found.");
            }
        }

        public static bool ServiceExists(string serviceName)
        {
            var service = LoadService(serviceName);

            if (service != null)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public static ServiceActionResult StopService(string serviceName)
        {
            var statusResult = ServiceActionResult.NotFound;

            if (ServiceExists(serviceName) == false)
            {
                return statusResult;
            }

            var service = LoadService(serviceName);

            if ((service != null && service.Status == ServiceControllerStatus.Running))
            {
                service.Stop();
                statusResult = ServiceActionResult.Stopped;
            }

            return statusResult;
        }

        public static ServiceActionResult StartService(string serviceName)
        {
            var statusResult = ServiceActionResult.NotFound;

            if (ServiceExists(serviceName) == false)
            {
                return statusResult;
            }

            var service = LoadService(serviceName);

            if ((service != null && service.Status == ServiceControllerStatus.Stopped))
            {
                service.Start();
                statusResult = ServiceActionResult.Running;
            }

            return statusResult;
        }

        public static IEnumerable<string> AllServices()
        {
            return ServiceController.GetServices().Select(p => p.ServiceName).AsEnumerable();
        }

        private static ServiceController LoadService(string serviceName)
        {
            return ServiceController.GetServices().Where(p => p.ServiceName == serviceName).FirstOrDefault();
        }

        public static void StartStopServices(IEnumerable<ServiceAction> requests)
        {
            Encapsulation.TryValidateParam(requests, nameof(requests));

            foreach (var request in requests)
            {
                if (request.ServiceActionRequest == ServiceActionRequest.Start)
                {
                    request.ServiceActionResult = StartService(request.ServiceName);
                }
                else if (request.ServiceActionRequest == ServiceActionRequest.Stop)
                {
                    request.ServiceActionResult = StopService(request.ServiceName);
                }
            }
        }
    }

    public enum ServiceActionResult
    {
        NotFound,
        Running,
        Stopped
    }

    public enum ServiceActionRequest
    {
        Unknown,
        Start,
        Stop
    }

    public class ServiceAction
    {
        public string ServiceName { get; set; }
        public ServiceActionRequest ServiceActionRequest { get; set; }
        public ServiceActionResult ServiceActionResult { get; set; }
    }
}
